feat(common): introduce KeyValuePipe

Toxicable opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[x] Feature
```

## What is the current behavior?
There is no way to iterate a object or Map

Closes: https://github.com/angular/angular/issues/24200
Continuation from: https://github.com/angular/angular/pull/11319

## What is the new behavior?
A pipe to object Map or object (dictionary) to a key value pair array for use in a ngFor

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

Commit:
feat(core): export defaultKeyValueDiffers to private api

Commit:
feat(common): introduce KeyValuePipe

mhevery commented on packages/common/src/pipes/entries_pipe.ts:
```diff
@@ -0,0 +1,70 @@
+  value: V;
+}
+
+@Pipe({name: 'mutableToKeyValue', pure: false})
```
I don' think the word `mutable` should be there. It is inconsistent with the rest of the systems naming.

Toxicable replied:
Changed to `keyvalue` similar to the `lowercase` on LowerCasePipe

mhevery commented on packages/common/src/pipes/entries_pipe.ts:
```diff
@@ -0,0 +1,70 @@
+      this.differ = this.differs.find(input).create();
+    }
+
+    // TODO: shouldnt the differ allow Map?
```
It does: https://github.com/angular/angular/blob/master/packages/core/src/change_detection/differs/default_keyvalue_differ.ts#L74 Not sure I understand that comment.

Toxicable replied:
I get a type error if I pass `input` directly

Toxicable replied:
specifically: 
```
Argument of type '{ [key: string]: V; [key: number]: V; } | Map<K, V>' is not assignable to parameter of type '{ [key: string]: V; }'.
  Type 'Map<K, V>' is not assignable to type '{ [key: string]: V; }'.
    Index signature is missing in type 'Map<K, V>'
```

mhevery commented on packages/common/src/pipes/entries_pipe.ts:
```diff
@@ -0,0 +1,70 @@
+    let nextValue: Array<KeyValuePair<string, V>>;
+    // always return a new array ref at this
+    if (input instanceof Map) {
+      // keys maintain their type here
```
I don't think we should be stringifing it here. What is your reasoning for it?

Toxicable replied:
Object.keys converts all keys to strings. So to be consistent we should do the same with Map

mhevery commented on packages/common/src/pipes/entries_pipe.ts:
```diff
@@ -0,0 +1,70 @@
+                      .map(key => this.makeKeyValuePair(String(key), input.get(key)));
+    } else {
+      // all keys will be converted into strings when using Object.keys()
+      nextValue = Object.keys(input).sort().map(key => this.makeKeyValuePair(key, input[key]));
```
`map` is not very performant. Given that this will be in the hotpath could we implement it with a for loop instead?

Toxicable replied:
Done.
Still not overly happy with `Array.from(input.keys())` since it's an extra iteration but I don't see another way to sort it without resolving the iterator into an Array

mhevery commented on packages/common/src/pipes/entries_pipe.ts:
```diff
@@ -0,0 +1,70 @@
+
+import {KeyValueDiffer, KeyValueDiffers, Pipe, PipeTransform} from '@angular/core';
+
+export interface KeyValuePair<K, V> {
```
I think `KeyValue` is shorter. I don't think the word `Pair` adds value.

mhevery commented on tools/public_api_guard/common/common.d.ts:
```diff
@@ -220,6 +220,14 @@ export declare class LowerCasePipe implements PipeTransform {
+    transform(input: null | {
+        [key: string]: V;
+        [key: number]: V;
+    } | Map<K, V>): Array<KeyValuePair<string, V>> | null;
```
I would expect that `KeyValuePair` should be exported as well since it would be part of the public API, but I don't see it in this file.

mhevery commented on packages/common/src/pipes/index.ts:
```diff
@@ -14,6 +14,7 @@
 import {AsyncPipe} from './async_pipe';
 import {LowerCasePipe, TitleCasePipe, UpperCasePipe} from './case_conversion_pipes';
 import {DatePipe} from './date_pipe';
+import {MutableToKeyValuePipe} from './entries_pipe';
```
I would expect `KeyValuePair` to be exported as well since it is part of public API.

Toxicable commented:
@mhevery I've addressed your feedback

mhevery commented on packages/common/src/pipes/entries_pipe.ts:
```diff
@@ -0,0 +1,77 @@
+}
+
+@Pipe({name: 'keyvalue', pure: false})
+export class KeyValuePipe<K, V> implements PipeTransform {
```
Sorry I did not notice this earlier. Because this is public API it needs to have documentation. See example here: https://github.com/angular/angular/blob/master/packages/common/src/pipes/date_pipe.ts#L14-L105

Toxicable replied:
Never been very good at docs, let me know if that'll suffice

mhevery commented on packages/common/src/pipes/entries_pipe.ts:
```diff
@@ -0,0 +1,77 @@
+    return nextValue;
+  }
+
+  private makeKeyValuePair(key: number|string, value: any): KeyValue<string, V> {
```
This method would be better as free floating function since it would minify better. (It does not use `this` hence it does not need to be a member of the class)

mhevery commented on packages/common/src/pipes/entries_pipe.ts:
```diff
@@ -0,0 +1,77 @@
+import {KeyValueDiffer, KeyValueDiffers, Pipe, PipeTransform} from '@angular/core';
+
+export interface KeyValue<K, V> {
+  key: K;
```
since `K` is always a `string` there is no need for `K`

mhevery commented on packages/common/src/pipes/entries_pipe.ts:
```diff
@@ -0,0 +1,77 @@
+    let nextValue: Array<KeyValue<string, V>> = [];
+
+    if (input instanceof Map) {
+      const keys = Array.from(input.keys()).sort();
```
If `keys` are not `string`s than the `sort` will fail. I think you need to convert to `string` before `sort`()

Toxicable replied:
Unless i've misinterpted you sort can handle `number` just fine.
take this example 
![image](https://user-images.githubusercontent.com/13490925/41078542-9118e2da-6a71-11e8-88d9-059c59169776.png)

Toxicable replied:
I think I get what you mean: when the input is a function, object etc.
I think adding a param for a sort comparitor function would be the best idea here. With the default behaviour being `.sort()` so leave it to the dev to add the comparitor if they have complex keys

mhevery commented on packages/common/src/pipes/entries_pipe.ts:
```diff
@@ -0,0 +1,77 @@
+      for (let i = 0; i < keys.length; i++) {
+        const key = keys[i];
+        // input could possibly be Megamorphic here
+        nextValue.push(this.makeKeyValuePair(String(key), input[key]));
```
No need to convert to `string` since `Object.keys` will always return `strings`

Toxicable replied:
Thanks, missed that in refactoring

mhevery reviewed (changes requested):
I am sorry I missed the docs the first time through. Hopefully this is the last pass.

alfaproject commented:
Question before this becomes public API: shouldn't the sort be optional? (maybe on by default, but configurable at least?)

mhevery commented on packages/common/src/pipes/keyvalue_pipe.ts:
```diff
@@ -0,0 +1,90 @@
+ * The output array will be ordered by keys.
+ * By default the comparator will be by Unicode point value
+ * You can optionally pass a compareFn if your keys are complex types.
+ *
```
Could you add an example of usage. We are trying to make sure that all new APIs have an example. See as an example of how to do it https://github.com/angular/angular/blob/master/packages/common/src/pipes/async_pipe.ts#L43-L68. Notice that the example https://github.com/angular/angular/blob/master/packages/examples/common/pipes/ts/async_pipe.ts has tests https://github.com/angular/angular/blob/master/packages/examples/common/pipes/ts/e2e_test/pipe_spec.ts which verifies that the example does not get broken.

mhevery commented on packages/common/src/pipes/keyvalue_pipe.ts:
```diff
@@ -0,0 +1,90 @@
+  private lastValue: Array<KeyValue<V>>;
+
+  transform(input: null|{
+    [key: string]: V;
```
replace `;` with `,`

mhevery commented on packages/common/src/pipes/keyvalue_pipe.ts:
```diff
@@ -0,0 +1,90 @@
+      this.differ = this.differs.find(input).create();
+    }
+
+    // TODO: shouldnt the differ signature allow Map?
```
`shouldnt` => `shouldn't`

The issue is that you declared your type as `{} | Map` where as `KeyValueDiffer` has: `  diff(object: Map<K, V>): KeyValueChanges<K, V>;` and `diff(object: {[key: string]: V}): KeyValueChanges<string, V>;` overloaded methods. The issue is that neither `{}` or `Map` is a subtype of `{}|Map`. To make this work you would have to declare `diff(object: {[key: string]: V}|Map<K,V> ): KeyValueChanges<K|string, V>;` or something like that, but than you would love the type information on `string`. 

So casting to `any` is a reasonable compromise. 

Alternatively you could do: 
```
    const differChanges = input instanceof Map ? this.differ.diff(input) : this.differ.diff(input);
```
Which types but is very weird.

mhevery commented on packages/common/src/pipes/keyvalue_pipe.ts:
```diff
@@ -0,0 +1,90 @@
+  return {key: key as any, value};
+}
+
+export interface KeyValue<V> {
```
KeyValue is part of public API, it needs docs. (example not needed)

mhevery commented:
OK, I thought about it some more and I have done some refactoring to clean up the code and the type system. I pushed the changes here https://github.com/angular/angular/pull/24319/commits/97fbe77331b1ddf354a341e819a074706df91571

Toxicable commented:
@mhevery Thanks man, that helps a lot, will make those final few docs and tests

alfaproject commented on packages/common/src/pipes/keyvalue_pipe.ts:
```diff
@@ -0,0 +1,104 @@
+ *
+ */
+@Pipe({name: 'keyvalue', pure: false})
+export class KeyValuePipe implements PipeTransform {
```
Any reason not to do `KeyValuePipe<K, V>`? It down-levels to the exact same code and would avoid having to use `any`, I think.

Commit:
feat(core): KeyValueDiffer#diff allows null values

Commit:
fixup! feat(common): introduce KeyValuePipe

Commit:
fixup! feat(common): introduce KeyValuePipe

Toxicable commented:
@mhevery Addressed all your feedback and added more tests / docs.
However there appears to be a failure on IE 10/11 which I don't really know about.
ref: https://travis-ci.org/angular/angular/jobs/391665522#L1237
which is this test here: https://github.com/angular/angular/pull/24319/files#diff-412c822ba885d9fcea7145c520161742R94
But I don't really see how any of the syntax or functions I'm using there wouldn't work in IE 10/11 after downleveling to es5

Commit:
fixup! feat(common): introduce KeyValuePipe

mhevery commented:
@Toxicable Thank you for all of your patience in getting this into mergeable state. 👍

mary-poppins commented:
You can preview d5bca8e at https://pr24319-d5bca8e.ngbuilds.io/.
You can preview 75cf173 at https://pr24319-75cf173.ngbuilds.io/.
You can preview d60066e at https://pr24319-d60066e.ngbuilds.io/.
You can preview 7fb7e8f at https://pr24319-7fb7e8f.ngbuilds.io/.
You can preview 617a924 at https://pr24319-617a924.ngbuilds.io/.
You can preview 70f8827 at https://pr24319-70f8827.ngbuilds.io/.
You can preview 6e25cdb at https://pr24319-6e25cdb.ngbuilds.io/.
You can preview 65f8ff6 at https://pr24319-65f8ff6.ngbuilds.io/.
You can preview 908cfa8 at https://pr24319-908cfa8.ngbuilds.io/.
You can preview 97fbe77 at https://pr24319-97fbe77.ngbuilds.io/.
You can preview 2aa846a at https://pr24319-2aa846a.ngbuilds.io/.
You can preview 4b60338 at https://pr24319-4b60338.ngbuilds.io/.
You can preview 851e923 at https://pr24319-851e923.ngbuilds.io/.
You can preview 351e25e at https://pr24319-351e25e.ngbuilds.io/.
You can preview 393b3a5 at https://pr24319-393b3a5.ngbuilds.io/.
You can preview b99eba1 at https://pr24319-b99eba1.ngbuilds.io/.
You can preview cfb35c7 at https://pr24319-cfb35c7.ngbuilds.io/.
You can preview 83ec620 at https://pr24319-83ec620.ngbuilds.io/.

mhevery commented:
http://test/OCL:200432210:BASE:200432373:1528917235420:35b9231e
